backend/src/main/scala/com/softwaremill/adopttapir/template/ProjectGenerator.scala

```diff
@@ -0,0 +1,212 @@
+      }
+
+    GeneratedFile(
+      pathUnderPackage("src/test/scala", groupId, "EndpointsSpec.scala"),
+      fileContent.toString
+    )
+  }
+
+  import CommonObjectTemplate.scalafmtConfigPath
+  private val scalafmtConf: ScalaVersion => GeneratedFile = dialectVersion =>
+    GeneratedFile(scalafmtConfigPath, txt.scalafmt(TemplateDependencyInfo.scalafmtVersion, dialectVersion).toString())
+
+  private def pathUnderPackage(prefixDir: String, groupId: String, fileName: String): String =
+    prefixDir + "/" + groupId.split('.').mkString("/") + "/" + fileName
+
+  private def toSortedList(set: Set[Import]): List[Import] = set.toList.sortBy(_.fullName)
+}
+
+object SbtProjectTemplate extends ProjectTemplate {
+  override def generate(starterDetails: StarterDetails): List[GeneratedFile] =
+    super.generate(starterDetails) ::: List(getBuildSbt(starterDetails), buildProperties, pluginsSbt, sbtx, readme)
+
+  lazy val sbtxFile = "sbtx"
+
+  private def getBuildSbt(starterDetails: StarterDetails): GeneratedFile = {
+    val content = txt
+      .sbtBuild(
+        starterDetails.projectName,
+        starterDetails.groupId,
+        starterDetails.scalaVersion.value,
+        TemplateDependencyInfo.tapirVersion,
+        (BuildSbtView.getAllDependencies _).andThen(BuildSbtView.format)(starterDetails),
+        starterDetails.serverEffect == ZIOEffect
+      )
+      .toString()
+
+    GeneratedFile("build.sbt", content)
+  }
+
+  private lazy val buildProperties: GeneratedFile = GeneratedFile(
+    "project/build.properties",
+    txt.buildProperties(TemplateDependencyInfo.sbtVersion).toString()
+  )
+
+  private lazy val pluginsSbt: GeneratedFile = GeneratedFile("project/plugins.sbt", CommonObjectTemplate.templateResource("plugins.sbt"))
+
+  private lazy val sbtx: GeneratedFile =
+    GeneratedFile(sbtxFile, CommonObjectTemplate.templateResource(sbtxFile))
+
+  private lazy val readme: GeneratedFile =
+    GeneratedFile(CommonObjectTemplate.readMePath, CommonObjectTemplate.templateResource("README_sbt.md"))
+}
+
+object CommonObjectTemplate {
+  def templateResource(fileName: String): String = Resource.getAsString(s"template/$fileName")
+
+  val scalafmtConfigPath = ".scalafmt.conf"
+
+  val readMePath: String = "README.md"
+
+  implicit class StarterDetailsWithLegalizedGroupId(starterDetails: StarterDetails) {
+    lazy val legalizedGroupId: String = legalizeGroupId(starterDetails)
+  }
+
+  private def legalizeGroupId(starterDetails: StarterDetails): String = {
+    def legalize(packageNameSection: String): String = {
+      val startsWithNumberRgx = "^\\d+.*$"
+
+      if (packageNameSection.matches(startsWithNumberRgx)) {
+        "_" + packageNameSection
+      } else {
+        packageNameSection
+      }
+    }
+
+    def legalizeGroupId(groupId: String): String = groupId.split('.').map(legalize).mkString(".")
+
+    legalizeGroupId(starterDetails.groupId)
+  }
+}
+
+private object ScalaCliProjectTemplate extends ProjectTemplate {
```

Comment:
[minor] I think instead of overriding methods I'd define methods such as `generateCode` and `generateBuild`, and then the `generate` impl would be `generateCode ++ generateBuild`